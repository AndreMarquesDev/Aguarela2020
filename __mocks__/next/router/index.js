jest.mock('next/router', () => ({
    ...jest.requireActual('next/router'),
    useRouter() {
        return {
            route: '/',
            pathname: '',
            query: {
                language: 'pt',
            },
            asPath: '',
        };
    },
}));

module.exports = jest.requireMock('next/router');
