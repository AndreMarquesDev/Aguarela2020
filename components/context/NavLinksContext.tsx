import { createContext, Context } from 'react';

export interface NavLinksContextProps {
    setNavHeight?: React.Dispatch<React.SetStateAction<number>>;
}

const NavLinksContext: Context<NavLinksContextProps> = createContext(null);

export default NavLinksContext;
