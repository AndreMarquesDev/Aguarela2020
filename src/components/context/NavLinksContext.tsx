import type { Context } from 'react';
import { createContext } from 'react';

export type NavLinksContextProps = {
    isMenuOpen: boolean;
    toggleMenu: React.Dispatch<React.SetStateAction<boolean>>;
    setNavHeight?: React.Dispatch<React.SetStateAction<number>>;
};

// @ts-ignore
export const NavLinksContext: Context<NavLinksContextProps> = createContext(null);
