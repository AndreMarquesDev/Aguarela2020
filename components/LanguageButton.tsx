import React, { FC } from 'react';
import { useRouter } from 'next/router';
import ReactCountryFlag from 'react-country-flag';
import { getRemainingLang, getPageFromUrl } from '../utils/routing/urlValidation';

interface ILanguageButtonProps {
    language: string;
}

const LanguageButton: FC<ILanguageButtonProps> = props => {
    const {
        language,
    } = props;

    const router = useRouter();
    const currentPage = getPageFromUrl[0];
    const languageToSwitchTo = getRemainingLang[0];
    const countryCode = language === 'en' ? 'PT' : 'GB';

    const onButtonClick = (): void => {
        // TODO: replace the correct url instead of redirecting to the homepage
        router.replace(`/${languageToSwitchTo}/${currentPage}`);
    };

    return (
        <>
            <button
                type="button"
                onClick={(): void => onButtonClick()}
            >
                <ReactCountryFlag
                    svg
                    countryCode={countryCode}
                    style={{
                        width: '30rem',
                        height: '30rem',
                    }}
                    title={countryCode}
                />
            </button>

            <style jsx>
                {`
                    button {
                        position: absolute;
                        right: 20rem;
                        bottom: 20rem;
                        transition: transform .2s;

                        &:hover {
                            transform: scale(.95);
                        }
                    }
                `}
            </style>
        </>
    );
};

export default LanguageButton;
