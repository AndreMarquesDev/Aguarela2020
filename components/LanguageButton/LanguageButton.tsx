import React, { useContext } from 'react';
import { useRouter } from 'next/router';
import { getRemainingLang } from 'multilingual-url/lib';
import { getPageFromUrl } from '../../utils/pages';
import { getCurrentLanguagetexts } from '../../utils/generic';
import { NavLinksContext } from '../context/NavLinksContext';
import { locales } from '../../constants/locales';
import { Locale } from '../../types/Locale';

export const LanguageButton = (): JSX.Element => {
    const router = useRouter();
    const currentPage = getPageFromUrl()[0];
    const languageToSwitchTo = getRemainingLang(locales)[0] as Locale;
    const translatedPageTitle: string = getCurrentLanguagetexts(languageToSwitchTo)[currentPage];

    const { isMenuOpen, toggleMenu } = useContext(NavLinksContext);

    const onButtonClick = (): void => {
        router.replace(`/${languageToSwitchTo}/${currentPage}`);
        document.title = `${translatedPageTitle} - Aguarela Digital`;

        if (isMenuOpen) {
            toggleMenu(!isMenuOpen);
        }
    };

    return (
        <>
            <button suppressHydrationWarning type="button" onClick={onButtonClick}>
                {languageToSwitchTo}
            </button>

            <style jsx>
                {`
                    @import './styles/_vars.scss';

                    button {
                        color: inherit;
                        text-transform: inherit;
                        opacity: 0.5;
                        outline: none;
                        transition: color 0.15s linear;
                    }
                `}
            </style>
        </>
    );
};
