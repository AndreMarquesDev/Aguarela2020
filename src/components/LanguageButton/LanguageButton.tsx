import type { Locale } from '../../types/Locale';
import type { TextsInterface } from '../../utils/texts';
import { getRemainingLang } from 'multilingual-url/lib';
import { useRouter } from 'next/router';
import React, { useContext } from 'react';
import { locales } from '../../constants/locales';
import { getCurrentLanguagetexts } from '../../utils/generic';
import { getPageFromUrl } from '../../utils/pages';
import { NavLinksContext } from '../context/NavLinksContext';

export const LanguageButton = (): JSX.Element => {
    const router = useRouter();
    const currentPage = getPageFromUrl()[0];
    const languageToSwitchTo = getRemainingLang(locales)[0] as Locale;
    const translatedPageTitle: string =
        getCurrentLanguagetexts(languageToSwitchTo)[currentPage as keyof TextsInterface];

    const { isMenuOpen, toggleMenu } = useContext(NavLinksContext);

    const onButtonClick = (): void => {
        router.replace(`/${languageToSwitchTo}/${currentPage}`);
        document.title = `${translatedPageTitle} - Aguarela Digital`;

        if (isMenuOpen) {
            toggleMenu(!isMenuOpen);
        }
    };

    return (
        <>
            <button suppressHydrationWarning type="button" onClick={onButtonClick}>
                {languageToSwitchTo}
            </button>

            <style jsx>
                {`
                    button {
                        color: inherit;
                        text-transform: inherit;
                        opacity: 0.5;
                        outline: none;
                        transition: color 0.15s linear;
                    }
                `}
            </style>
        </>
    );
};
