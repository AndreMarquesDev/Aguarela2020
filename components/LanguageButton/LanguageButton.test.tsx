import '@testing-library/jest-dom';
import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as nextRouter from 'next/router';
import { LanguageButton } from './LanguageButton';
import NavLinksContext from '../context/NavLinksContext';

// @ts-ignore
nextRouter.useRouter = jest.fn(() => ({
    query: {
        language: 'pt',
    },
    replace: jest.fn(),
}));

describe('<LanguageButton />', () => {
    test('renders properly', () => {
        const { container } = render(
            <NavLinksContext.Provider value={{ isMenuOpen: false, toggleMenu: jest.fn() }}>
                <LanguageButton />
            </NavLinksContext.Provider>
        );

        const button = screen.getByRole('button');

        expect(button).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    test('renders properly when clicking on button', () => {
        Object.defineProperty(window, 'location', {
            value: {
                pathname: '/en/about',
            },
            writable: true,
        });

        render(
            <NavLinksContext.Provider value={{ isMenuOpen: false, toggleMenu: jest.fn() }}>
                <LanguageButton />
            </NavLinksContext.Provider>
        );

        const button = screen.getByText('pt');

        expect(button).toBeInTheDocument();

        userEvent.click(button);

        expect(document.title).toBe('Sobre - Aguarela Digital');
    });
});
