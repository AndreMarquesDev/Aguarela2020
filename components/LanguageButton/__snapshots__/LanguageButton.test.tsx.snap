// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LanguageButton /> renders properly 1`] = `
<div>
  <button
    class="jsx-1698987686"
    type="button"
  >
    pt
  </button>
</div>
`;
